### DISCUSSION TITLE
Coffee-loader failed to process CoffeeScript files with CoffeeScriptError: error: reserved word 'var'

### DISCUSSION THREAD

**Discussion by talalong** [webpack/webpack#18513] [Q&A] [answered]
I am trying to upgrade the Webpacker for our legacy Rails web application, which was using Webpacker V5. The goal is to upgrade from Webpacker V5 to V6 (6.0.0-rc.6) and then move to Shakapacker. However, I am stuck at coffee-loader. It seems that the CoffeeScript files are not being converted correctly.

Running `rake assets:precompile` leads to an error. I created a simple CoffeeScript file to test it, and this is the error I got:

(Note: The old Webpacker V5 configuration worked fine.)

I have tried switching coffee-loader from V1 to V5, but it didn’t help. What am I doing wrong here?

>  Module build failed (from ./node_modules/coffee-loader/dist/cjs.js):
    error: reserved word 'var'
    var hello;
    ^^^

This is my dummy coffeescript file:
```
hello = -> console.log('hi!')
```

This is my webpacker config:

```
//config/webpack/environment.js

const { webpackConfig } = require('@rails/webpacker')
const { VueLoaderPlugin } = require('vue-loader')
const vue =  require('./loaders/vue')
const coffee = require('./loaders/coffee')
const sass = require('./loaders/sass.js')
const yaml = require('./loaders/yaml.js')
const pug = require('./loaders/pug.js')
const webpack = require('webpack');

webpackConfig.plugins = [
    ...webpackConfig.plugins,
    new VueLoaderPlugin(),
    new webpack.DefinePlugin({
        __VUE_OPTIONS_API__: true,
        __VUE_PROD_DEVTOOLS__: false,
    })
]

webpackConfig.module.rules = [
    vue,
    ...webpackConfig.module.rules,
    pug,
    coffee,
    sass,
    yaml
]

delete webpackConfig.module.rules['nodeModules']

const scssConfigIndex = webpackConfig.module.rules.findIndex((config) => ".scss".match(config.test))
webpackConfig.module.rules.splice(scssConfigIndex, 1)

webpackConfig.resolve = {
  extensions: [".coffee", ".vue", ".mjs", ".js", ".sass", ".scss", ".css", ".module.sass",
    ".module.scss", ".module.css", ".png", ".svg", ".gif", ".jpeg", ".jpg"]
}

module.exports = webpackConfig

```

`loaders/coffee.js`:

```
module.exports = {
  test: /\.coffee$/,
  use: ["babel-loader", "coffee-loader"]
}
```

Here is the `package.json`:

```
{
  "dependencies": {
    "@rails/ujs": "^7.1.3",
    "@rails/webpacker": "6.0.0-rc.6",
    "axios": "^0.28.1",
    "coffee-loader": "^2.0.0",
    "coffeescript": "^2.7.0",
    "css-loader": "^7.1.2",
    "date-fns": "^2.30.0",
    "file-loader": "^6.2.0",
    "file-saver": "^2.0.5",
    "foundation-datepicker": "^1.5.6",
    "foundation-icons": "^2.0.0",
    "lodash.mapkeys": "^4.6.0",
    "lodash.pick": "^4.4.0",
    "lodash.pickby": "^4.6.0",
    "micromodal": "^0.4.10",
    "moment": "^2.30.1",
    "node-sass": "^9.0.0",
    "papaparse": "^5.4.1",
    "pug": "^3.0.3",
    "pug-plain-loader": "^1.1.0",
    "resolve-url-loader": "^5.0.0",
    "sass-loader": "^14.2.1",
    "vue": "^3.4.29",
    "vue-axios": "^3.5.2",
    "vue-content-loader": "^2.0.0",
    "vue-easy-spinner": "^0.1.0",
    "vue-final-modal": "^3.4.11",
    "vue-i18n": "9.2.2",
    "vue-loader": "^17.4.2",
    "vue-loading-overlay": "5.0.3",
    "vue-multiselect": "^3.0.0",
    "vue-router": "^4.3.2",
    "vue-style-loader": "^4.1.3",
    "vue-turbolinks": "^2.2.2",
    "vue3-datepicker": "^0.2.5",
    "vuex": "^4.1.0",
    "webpack": "^5.91.0",
    "webpack-cli": "^5.1.4",
    "yaml-js-loader": "^0.1.1"
  },
  "devDependencies": {
    "@babel/core": "^7.24.7",
    "@babel/plugin-proposal-class-properties": "^7.18.6",
    "@babel/plugin-proposal-object-rest-spread": "^7.20.7",
    "@babel/plugin-proposal-private-methods": "^7.18.6",
    "@babel/plugin-proposal-private-property-in-object": "^7.21.11",
    "@babel/plugin-transform-runtime": "^7.24.7",
    "@babel/preset-env": "^7.24.7",
    "@babel/runtime": "^7.24.7",
    "@coffeelint/cli": "^5.2.11",
    "@vue/compiler-sfc": "^3.3.4",
    "@vue/test-utils": "^2.4.1",
    "@vue/vue3-jest": "^27.0.0",
    "@webpack-cli/serve": "^2.0.5",
    "babel-core": "^7.0.0-bridge.0",
    "babel-jest": "^27.5.1",
    "babel-loader": "^9.1.3",
    "babel-plugin-macros": "^3.1.0",
    "clean-webpack-plugin": "4.0.0",
    "flush-promises": "^1.0.2",
    "jest": "^27.2.5",
    "jsverify": "^0.8.4",
    "mini-css-extract-plugin": "^2.9.0",
    "rosie": "^2.1.1",
    "webpack-dev-server": "^4.15.1"
  },
  "browserslist": [
    "defaults"
  ],
  "babel": {
    "presets": [
      "./node_modules/@rails/webpacker/package/babel/preset.js"
    ]
  },
  "scripts": {
    "test": "TZ=UTC jest",
    "coffeelint": "find app -name '*.coffee' | xargs coffeelint"
  },
  "jest": {
    "testEnvironment": "jsdom",
    "moduleFileExtensions": [
      "js",
      "json",
      "vue",
      "coffee"
    ],
    "roots": [
      "app/javascript"
    ],
    "transform": {
      "\\.coffee$": "<rootDir>/config/jest-coffee-preprocessor/index.js",
      "^.+\\.vue$": "@vue/vue3-jest",
      "^.+\\.js$": "<rootDir>/node_modules/babel-jest"
    },
    "globals": {
      "vue-jest": {
        "pug": {
          "doctype": "html"
        }
      }
    },
    "transformIgnorePatterns": [
      "/node_modules/(?!vue-multiselect)"
    ],
    "setupFilesAfterEnv": [
      "<rootDir>/config/jest/pre-test-setup.js"
    ],
    "collectCoverage": true,
    "collectCoverageFrom": [
      "**/javascript/**/*.{js,vue,coffee}"
    ],
    "coveragePathIgnorePatterns": [
      "packs/*",
      "apps/*"
    ],
    "coverageDirectory": "jest-coverage",
    "testRegex": "\\.spec\\.(js|coffee)$"
  },
  "packageManager": "yarn@3.1.0"
}

```

**Comment by alexander-akait** [comment]
Weird, sounds like you can't use `var` in `coffee-loader`, will do it work before? Maybe you can create a minimum reproducible test repo?

**Reply by talalong** [reply]
> will do it work before

Yes. as it was using Webpacker V5 and with the old configuration syntax like `environment.loaders.append('coffee', coffee)`.

> Maybe you can create a minimum reproducible test repo?

I did, but I failed to reproduce the issue 😔. The CoffeeScript file gets converted to JS without me defining the coffee-loader. I don’t understand.

Here is the sample coffee:

```
hello = -> console.log('Hiiiiiii')

greet = (name) ->
  console.log "Hello, #{name}!"
  
greet "Alice"
```

And here is the output of `rake assets:precompile`
**public/assets/packs**

```
(function() {
  var greet, hello;

  hello = function() {
    return console.log('Hiiiiiii');
  };

  greet = function(name) {
    return console.log("Hello, " + name + "!");
  };

  greet("Alice");

}).call(this);

```

Here is my sample [repo](https://github.com/talalong/mmog-base), if you want to check.

**Reply by alexander-akait** [reply]
Maybe the new version of `Webpacker` already has coffee-loader by default

**Reply by talalong** [reply]
@alexander-akait Nope. That was just me forgetting to remove Sprocket from my Gemfile. By removing Sprocket and its configuration, I was able to reproduce the issue. And the coffee-loader is needed. Please check the file [`base.js`](https://github.com/talalong/mmog-base/blob/main/config/webpack/base.js)

> ERROR in ./app/javascript/packs/main.coffee
Module build failed (from ./node_modules/coffee-loader/dist/cjs.js):
CoffeeScriptError: /Users/long/Documents/projects/mmog-base/app/javascript/packs/main.coffee:1:1: error: reserved word 'var'
var huhuhu;
^^^

**Reply by talalong** [reply, marked as answer]
@alexander-akait  think the issue might be related to Webpacker V6 and the coffee-loader. I have created another test repo for Webpacker V5 (pls see [this](https://github.com/talalong/rails-packv5)). It has the same setup as I have in V6, and there is no problem there (except `coffee-script` needs to remain at v1).

**Reply by alexander-akait** [reply]
Sorry for delay, did you solve the problem?

**Reply by talalong** [reply]
@alexander-akait yes. I use shakapacker now as a replacement for the webpacker and I dont have the issue anymore. Thank you.

**Reply by jolll4** [reply]
What was the issue in the end? I'm doing a similar migration of V5->V6->shaka and coffee is the final thing that doesn't work as it worked before.

**Reply by talalong** [reply]
@jolll4 For me, the issue was that Webpacker V6 was expecting a different version of `coffee-script`. So, I skipped version V6 and went straight to Shaka V6. One thing I noticed is that Shaka already has the `coffee-script` loader installed, so I had to remove the coffee-loader from my old configuration rule.
